### PR TITLE
wildmidi: add new recipe, version 0.4.5

### PR DESCRIPTION
**Pull request by dbarkar**
Specify library name and version:  **wildmidi/0.4.5**

WildMIDI is a simple software midi player which has a core softsynth library that can be used with other applications.

Unlike other MIDI libraries here, it supports Redbook's extended MIDI files (XMI) that are used by many old games.

There are few issues that I'd like to resolve, I don't know where else I can ask for help.

- On Windows, CMAKE_BUILD_TYPE is not set automatically from build_type, so I had to set it explicitly. Why is this happening?
- CMake's find_library can't find "libWildMidi" without ".a"/."so" extension on Linux, so I had to add it in conanfile.py. Why? Why is it working for other recipes?
- Sometimes I get this error
`[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component wildmidi::wildmidi library 'libwildmidi' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054)`
Even though the library is in correct location, test_package and my app works well. Is it ok?
- There are files in lib/cmake describing WildMidi::libwildmidi target but this lines produce vague errors
`self.cpp_info.set_property("cmake_file_name", "wildmidi")`
`self.cpp_info.set_property("cmake_target_name", "libwildmidi")`
How to debug this issue?

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/wildmidi/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+
+required_conan_version = ">=1.53.0"
+
+
+class WildmidiConan(ConanFile):
+    name = "wildmidi"
+    version = "0.4.5"
+    description = "The WildMIDI library"
+    license = ("GPL-3.0+", "LGPL-3.0")
```
Comment (jwillikers) [inline recipes/wildmidi/all/conanfile.py:15]: For the licenses, can you use the [SPDX identifiers](https://spdx.org/licenses/)?

**Review by jwillikers — CHANGES_REQUESTED** [review]
Looks really good so far!

File: recipes/wildmidi/all/conanfile.py (hunk 1 above)
Comment (dbarkar) [reply recipes/wildmidi/all/conanfile.py:15]: I thought I'm using them, there are "GPL-3.0+" and "LGPL-3.0" identifiers in the table. If I'm missing something, can you elaborate?

Comment (jwillikers) [reply recipes/wildmidi/all/conanfile.py:15]: Yes, I wasn't very clear. The `GPL-3.0+` and `LGPL-3.0` show up, but only in the deprecated table. You have to scroll up to the table heading to see that it's the "deprecated" identifiers list.

The updated identifiers would be `GPL-3.0-or-later` and `LGPL-3.0-only`.

However, since this recipe is only packaging the library and not the player, it only falls under the LGPL, so it should be fine to drop the GPL one unless you want to add an option to the recipe to build the player. In such a case, you can put _both_ licenses in the license attribute and set the license appropriately in the `configure` method.

You should probably handle the licenses that are copied in via the `package` method similarly.

Comment (dbarkar) [reply recipes/wildmidi/all/conanfile.py:15]: It's all clear now, thanks! I'm not packaging the player, so only LGPL license is packaged and set in conanfile

**Comment by jwillikers** [comment]
> * On Windows, CMAKE_BUILD_TYPE is not set automatically from build_type, so I had to set it explicitly. Why is this happening?

On Windows, Conan Center's CI is building with the Visual Studio generator, which is a multi-configuration generator. Multiconfiguration generators do not set or use `CMAKE_BUILD_TYPE`. WildMIDI's top-level CMakeLists.txt needs to take this into account when setting things based on the build type. Generator expressions are used for the purpose of setting options based on the build type since they work for multi-configuration and single-configuration generators.
Also, WildMIDI sets the `CMAKE_CONFIGURATION_TYPES` variable, which is not a safe thing to do.

> * CMake's find_library can't find "libWildMidi" without ".a"/."so" extension on Linux, so I had to add it in conanfile.py. Why? Why is it working for other recipes?

It probably adds the `lib` prefix along with the possible suffixes. Maybe try setting the library name to `WildMidi` instead.

> * Sometimes I get this error
>   `[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component wildmidi::wildmidi library 'libwildmidi' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054)`
>   Even though the library is in correct location, test_package and my app works well. Is it ok?

Try switching the library name to `WildMidi` and that might fix this.

> * There are files in lib/cmake describing WildMidi::libwildmidi target but this lines produce vague errors
>   `self.cpp_info.set_property("cmake_file_name", "wildmidi")`
>   `self.cpp_info.set_property("cmake_target_name", "libwildmidi")`
>   How to debug this issue?

I suspect that these will work if you rename the library, but the `cmake_target_name` should be the full target name with the `::`. In the new CMakeDeps generator, the file name corresponds to the name used to find the library via `find_package`, i.e. `WildMidi` in this case since the file name is `WildMidiConfig.cmake`.

```
self.cpp_info.set_property("cmake_file_name", "WildMidi")
self.cpp_info.set_property("cmake_target_name", "WildMidi::libwildmidi")
```

**Comment by dbarkar** [comment]
Thanks! All issues are resolved now :)
Changing library name to WildMidi worked! It still has to be libWildMidi for MSVC, though, the code is much cleaner now.
This code seemed to work:
```
self.cpp_info.set_property("cmake_file_name", "wildmidi")
self.cpp_info.set_property("cmake_target_name", "wildmidi::wildmidi")
```
I not a CMake expert, I don't really know what I'm doing, please check if it's ok :)

File: recipes/wildmidi/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+
+class WildmidiConan(ConanFile):
+    name = "wildmidi"
+    version = "0.4.5"
+    description = "WildMIDI is a simple software midi player which has a core softsynth library that can be used in other applications."
+    license = "LGPL-3.0-only"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://www.mindwerks.net/projects/wildmidi"
+    topics = ("midi", "audio", "sound", "multimedia")
```
Comment (jwillikers) [inline recipes/wildmidi/all/conanfile.py:18]: I thing it would help others find this library by including the tags `softsynth`, `synth`, and `music`.

```suggestion
    topics = ("audio", "midi", "multimedia", "music", "softsynth", "sound", "synth")
```

File: recipes/wildmidi/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        self.cpp_info.libs = [libname]
+
+        if not self.options.shared:
+            self.cpp_info.defines = ["WILDMIDI_STATIC"]
+
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs.append("m")
+
+        self.cpp_info.set_property("cmake_file_name", "wildmidi")
+        self.cpp_info.set_property("cmake_target_name", "wildmidi::wildmidi")
```
Comment (jwillikers) [inline recipes/wildmidi/all/conanfile.py:89]: We just need to tweak the names of the provided files here to match the ones provided by the project itself.

```suggestion
        self.cpp_info.set_property("cmake_file_name", "WildMidi")
        self.cpp_info.set_property("cmake_target_name", "WildMidi::libwildmidi")
```

File: recipes/wildmidi/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        self.cpp_info.set_property("pkg_config_name", "wildmidi")
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = "wildmidi"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "wildmidi"
+        self.cpp_info.names["cmake_find_package"] = "wildmidi"
+        self.cpp_info.names["cmake_find_package_multi"] = "wildmidi"
```
Comment (jwillikers) [inline recipes/wildmidi/all/conanfile.py:96]:
```suggestion
        self.cpp_info.filenames["cmake_find_package"] = "WildMidi"
        self.cpp_info.filenames["cmake_find_package_multi"] = "WildMidi"
        self.cpp_info.names["cmake_find_package"] = "libwildmidi"
        self.cpp_info.names["cmake_find_package_multi"] = "libwildmidi"
```

File: recipes/wildmidi/all/test_package/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.8)
+
+project(test_package C)
+
+find_package(wildmidi REQUIRED CONFIG)
```
Comment (jwillikers) [inline recipes/wildmidi/all/test_package/CMakeLists.txt:5]:
```suggestion
find_package(WildMidi REQUIRED CONFIG)
```

File: recipes/wildmidi/all/test_package/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+
+add_executable(${PROJECT_NAME} test_package.c)
+target_link_libraries(${PROJECT_NAME} PRIVATE wildmidi::wildmidi)
```
Comment (jwillikers) [inline recipes/wildmidi/all/test_package/CMakeLists.txt:8]:
```
target_link_libraries(${PROJECT_NAME} PRIVATE WildMidi::libwildmidi)
```

**Review by jwillikers — CHANGES_REQUESTED** [review]
Just need to tweak the CMake target names and config filename to match those provided by the upstream project.

**Comment by dbarkar** [comment]
So with latest changes, I have build errors:
```
wildmidi/0.4.5 (test package): CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/conan/project/test_package/build/Release/generators/conan_toolchain.cmake" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/conan/project/test_package/."
-- Using Conan toolchain: /home/conan/project/test_package/build/Release/generators/conan_toolchain.cmake
-- Conan: Target declared 'WildMidi::libwildmidi'
-- Configuring done
-- Generating done
-- Build files have been written to: /home/conan/project/test_package/build/Release
wildmidi/0.4.5 (test package): CMake command: cmake --build "/home/conan/project/test_package/build/Release" '--' '-j28'
Scanning dependencies of target test_package
[ 50%] Building C object CMakeFiles/test_package.dir/test_package.c.o
/home/conan/project/test_package/test_package.c:4:10: fatal error: wildmidi_lib.h: No such file or directory
    4 | #include <wildmidi_lib.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
```
Test package can't find include file.
Specified in `/home/conan/project/test_package/build/Release/generators/conan_toolchain.cmake`:
```
list(PREPEND CMAKE_INCLUDE_PATH "/home/conan/.conan/data/wildmidi/0.4.5/_/_/package/dfbe50feef7f3c6223a476cd5aeadb687084a646/include")
```
I checked that dir and it's right there.
Also there is another error from CI.
I wonder why is it happening? I only changed the name.
Windows build works great.

**Review by jwillikers — APPROVED** [review]
Awesome work! Thanks for the contribution!

File: recipes/wildmidi/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+        self.cpp_info.set_property("cmake_file_name", "WildMidi")
+        self.cpp_info.set_property("cmake_target_name", "WildMidi::libwildmidi")
+        self.cpp_info.set_property("pkg_config_name", "WildMidi")
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = "WildMidi"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "WildMidi"
+        self.cpp_info.names["cmake_find_package"] = "libwildmidi"
+        self.cpp_info.names["cmake_find_package_multi"] = "libwildmidi"
```
Comment (jwillikers) [inline recipes/wildmidi/all/conanfile.py:96]:
```suggestion
        self.cpp_info.set_property("cmake_file_name", "WildMidi")
        self.cpp_info.set_property("cmake_target_name", "WildMidi::libwildmidi")
        self.cpp_info.set_property("pkg_config_name", "wildmidi")

        # TODO: back to global scope in conan v2 once cmake_find_package* generators removed
        self.cpp_info.components["libwildmidi"].libs = [libname]
        if not self.options.shared:
            self.cpp_info.defines = ["WILDMIDI_STATIC"]
        if self.settings.os in ["Linux", "FreeBSD"]:
            self.cpp_info.components["libwildmidi"].system_libs.append("m")

        # TODO: remove in conan v2 once cmake_find_package* generators removed
        self.cpp_info.names["cmake_find_package"] = "WildMidi"
        self.cpp_info.names["cmake_find_package_multi"] = "WildMidi"
        self.cpp_info.components["libwildmidi"].names["cmake_find_package"] = "libwildmidi"
        self.cpp_info.components["libwildmidi"].names["cmake_find_package_multi"] = "libwildmidi"
```

**Review by jwillikers — CHANGES_REQUESTED** [review]

**Comment by dbarkar** [comment]
Thanks for help!

**Review by jwillikers — DISMISSED** [review]

File: recipes/wildmidi/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+import os
+
+
+required_conan_version = ">=1.53.0"
+
+
+class WildmidiConan(ConanFile):
+    name = "wildmidi"
+    version = "0.4.5"
```
Comment (uilianries) [inline recipes/wildmidi/all/conanfile.py:13]:
```suggestion
```
Please, do not add version attribute, use conandata.yml instead, so we can re-use the same recipe for any future version.

**Review by uilianries — COMMENTED** [review]
Looking good, just a small comment.

**Review by jwillikers — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

File: recipes/wildmidi/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        if self.settings.os == "Windows":
```
Comment (prince-chrismc) [inline recipes/wildmidi/all/conanfile.py:49]: @jcar87 the behavior of CMakeToolchain is a little wonky here if you can take a look please

File: recipes/wildmidi/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            tc.variables["WANT_STATIC"] = True
+        tc.generate()
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def package(self):
+        copy(self, pattern="docs/license/LGPLv3.txt", dst=os.path.join(
```
Comment (prince-chrismc) [inline recipes/wildmidi/all/conanfile.py:62]: The docs says there's also a GPL licenseny reason why is that not copied?

Comment (prince-chrismc) [reply recipes/wildmidi/all/conanfile.py:62]: Nvm I found its only the demo app

**Review by prince-chrismc — APPROVED** [review]

File: recipes/wildmidi/all/conanfile.py (hunk 9 above)
Comment (jwillikers) [reply recipes/wildmidi/all/conanfile.py:49]: I think this is more a problem with the project than CMakeToolchain. The project seems to incorrectly handle multi-configuration generators, like the Visual Studio generator.

Comment (prince-chrismc) [reply recipes/wildmidi/all/conanfile.py:49]: Absolutely agreed but I am curious if this is the right solution of perhaps patches might be more robust... maybe fixing upstream could be an option?

Comment (dbarkar) [reply recipes/wildmidi/all/conanfile.py:49]: So the right thing to do is to add a proper support for multi-config. I will try to make a change in upstream and come back with new PR when new version comes out.

Comment (prince-chrismc) [reply recipes/wildmidi/all/conanfile.py:49]: I definitely think upstream could be better but I also think this is an edge case the new Conan generators weren't tested with so it would be nice to at least have test code to make sure this is intended 

So I want to be clear, this PR is good :)
